File: src/lib/db/repositories/setRepository.ts
import { and, eq, inArray, isNull, sql } from "drizzle-orm";

import type { ConceptTagId } from "@/data/conceptTags";
import type { AppDatabase } from "@/lib/db/client";
import {
  setItems,
  setTags,
  sets,
  variantSetAttachments,
  buildVariants,
  builds,
} from "@/lib/db/schema";
import type { SetType } from "@/lib/sets/schemas";

export type SetRecord = {
  id: string;
  userId: number;
  name: string;
  type: SetType;
  tagIds: ConceptTagId[];
  createdAt: string;
  updatedAt: string;
};

export type SetAttachmentRef = {
  buildId: string;
  buildName: string;
  variantId: string;
  variantName: string;
};

function rowToSet(row: typeof sets.$inferSelect, tagIds: ConceptTagId[]): SetRecord {
  return {
    id: row.id,
    userId: row.userId,
    name: row.name,
    type: row.type as SetType,
    tagIds,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  };
}

function loadTagsForSet(db: AppDatabase, setId: string): ConceptTagId[] {
  return db
    .select({ tagId: setTags.tagId })
    .from(setTags)
    .where(eq(setTags.setId, setId))
    .all()
    .map((r) => r.tagId as ConceptTagId)
    .sort();
}

export function listSets(db: AppDatabase, userId: number, type?: SetType): SetRecord[] {
  const rows = db
    .select()
    .from(sets)
    .where(
      type
        ? and(eq(sets.userId, userId), eq(sets.type, type))
        : eq(sets.userId, userId),
    )
    .all();

  return rows.map((row) => rowToSet(row, loadTagsForSet(db, row.id)));
}

export function listSetsByTags(
  db: AppDatabase,
  userId: number,
  tagIds: ConceptTagId[],
  type?: SetType,
): SetRecord[] {
  if (tagIds.length === 0) return listSets(db, userId, type);

  let matching: string[] | null = null;
  for (const tagId of tagIds) {
    const ids = db
      .select({ setId: setTags.setId })
      .from(setTags)
      .innerJoin(sets, eq(setTags.setId, sets.id))
      .where(and(eq(setTags.tagId, tagId), eq(sets.userId, userId)))
      .all()
      .map((r) => r.setId);

    matching = matching === null ? ids : matching.filter((id) => ids.includes(id));
  }

  if (!matching || matching.length === 0) return [];

  const rows = db
    .select()
    .from(sets)
    .where(
      type
        ? and(eq(sets.userId, userId), eq(sets.type, type), inArray(sets.id, matching))
        : and(eq(sets.userId, userId), inArray(sets.id, matching)),
    )
    .all();

  return rows.map((row) => rowToSet(row, loadTagsForSet(db, row.id)));
}

export function getSet(db: AppDatabase, userId: number, id: string): SetRecord | null {
  const row = db
    .select()
    .from(sets)
    .where(and(eq(sets.id, id), eq(sets.userId, userId)))
    .get();
  if (!row) return null;
  return rowToSet(row, loadTagsForSet(db, row.id));
}

export function createSetRecord(
  db: AppDatabase,
  userId: number,
  input: { id: string; name: string; type: SetType; tagIds: ConceptTagId[]; now: string },
): SetRecord {
  db.insert(sets)
    .values({
      id: input.id,
      userId,
      name: input.name,
      type: input.type,
      createdAt: input.now,
      updatedAt: input.now,
    })
    .run();

  for (const tagId of input.tagIds) {
    db.insert(setTags).values({ setId: input.id, tagId }).run();
  }

  return {
    id: input.id,
    userId,
    name: input.name,
    type: input.type,
    tagIds: input.tagIds,
    createdAt: input.now,
    updatedAt: input.now,
  };
}

export function updateSetRecord(
  db: AppDatabase,
  userId: number,
  id: string,
  patch: { name?: string; type?: SetType; tagIds?: ConceptTagId[]; now: string },
): SetRecord | null {
  const existing = getSet(db, userId, id);
  if (!existing) return null;

  db.update(sets)
    .set({
      name: patch.name ?? existing.name,
      type: patch.type ?? existing.type,
      updatedAt: patch.now,
    })
    .where(and(eq(sets.id, id), eq(sets.userId, userId)))
    .run();

  if (patch.tagIds) {
    db.delete(setTags).where(eq(setTags.setId, id)).run();
    for (const tagId of patch.tagIds) {
      db.insert(setTags).values({ setId: id, tagId }).run();
    }
  }

  return getSet(db, userId, id);
}

export function deleteSetRecord(db: AppDatabase, userId: number, id: string): boolean {
  const result = db
    .delete(sets)
    .where(and(eq(sets.id, id), eq(sets.userId, userId)))
    .run();
  return result.changes > 0;
}

export function findDuplicateName(
  db: AppDatabase,
  userId: number,
  type: SetType,
  name: string,
  excludeId?: string,
): boolean {
  const row = db
    .select({ id: sets.id })
    .from(sets)
    .where(and(eq(sets.userId, userId), eq(sets.type, type), eq(sets.name, name)))
    .get();
  if (!row) return false;
  if (excludeId && row.id === excludeId) return false;
  return true;
}

export function findAttachmentsBySetId(db: AppDatabase, setId: string): SetAttachmentRef[] {
  return db
    .select({
      buildId: builds.id,
      buildName: builds.name,
      variantId: buildVariants.id,
      variantName: buildVariants.name,
    })
    .from(variantSetAttachments)
    .innerJoin(buildVariants, eq(variantSetAttachments.variantId, buildVariants.id))
    .innerJoin(builds, eq(buildVariants.buildId, builds.id))
    .where(eq(variantSetAttachments.setId, setId))
    .all();
}

export function countActiveItemsInSet(db: AppDatabase, setId: string): number {
  const row = db
    .select({ count: sql<number>`count(*)` })
    .from(setItems)
    .where(and(eq(setItems.setId, setId), isNull(setItems.removedAt)))
    .get();
  return row?.count ?? 0;
}

File: packages/db/lib/src/repos/set_repository.dart
import 'package:drift/drift.dart';

import '../app_database.dart';
import '../tables.dart';
import 'library_records.dart';

SetRecord _rowToSet(LibrarySet row, List<String> tagIds) {
  return SetRecord(
    id: row.id,
    userId: row.userId,
    name: row.name,
    type: row.type,
    tagIds: tagIds,
    optimizerConstraints: row.optimizerConstraints,
    linkedModSetId: row.linkedModSetId,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  );
}

Future<List<String>> _loadTagsForSet(AppDatabase db, String setId) async {
  final rows = await (db.select(db.setTags)
        ..where((t) => t.setId.equals(setId)))
      .get();
  final tags = rows.map((r) => r.tagId).toList()..sort();
  return tags;
}

Future<Map<String, List<String>>> _loadTagsForSetIds(
  AppDatabase db,
  List<String> setIds,
) async {
  final map = {for (final id in setIds) id: <String>[]};
  if (setIds.isEmpty) return map;
  final rows = await (db.select(db.setTags)
        ..where((t) => t.setId.isIn(setIds)))
      .get();
  for (final row in rows) {
    map.putIfAbsent(row.setId, () => <String>[]).add(row.tagId);
  }
  for (final id in map.keys) {
    map[id]!.sort();
  }
  return map;
}

Future<List<SetRecord>> _rowsToSets(
  AppDatabase db,
  List<LibrarySet> rows,
) async {
  if (rows.isEmpty) return const [];
  final tagsBySet = await _loadTagsForSetIds(
    db,
    rows.map((r) => r.id).toList(),
  );
  return rows
      .map((row) => _rowToSet(row, tagsBySet[row.id] ?? const []))
      .toList();
}

/// List sets for [userId], optionally filtered by [type].
Future<List<SetRecord>> listSets(
  AppDatabase db,
  int userId, {
  String? type,
}) async {
  final query = db.select(db.sets)..where((t) => t.userId.equals(userId));
  if (type != null) {
    query.where((t) => t.type.equals(type));
  }
  final rows = await query.get();
  return _rowsToSets(db, rows);
}

Future<SetRecord?> getSet(AppDatabase db, int userId, String id) async {
  final row = await (db.select(db.sets)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .getSingleOrNull();
  if (row == null) return null;
  return _rowToSet(row, await _loadTagsForSet(db, row.id));
}

Future<SetRecord> createSetRecord(
  AppDatabase db,
  int userId, {
  required String id,
  required String name,
  required String type,
  List<String> tagIds = const [],
  String? optimizerConstraints,
  String? linkedModSetId,
  required String now,
}) async {
  await db.into(db.sets).insert(
        SetsCompanion.insert(
          id: id,
          userId: userId,
          name: name,
          type: type,
          optimizerConstraints: Value(optimizerConstraints),
          linkedModSetId: Value(linkedModSetId),
          createdAt: now,
          updatedAt: now,
        ),
      );
  for (final tagId in tagIds) {
    await db
        .into(db.setTags)
        .insert(SetTagsCompanion.insert(setId: id, tagId: tagId));
  }
  return (await getSet(db, userId, id))!;
}

Future<SetRecord?> updateSetRecord(
  AppDatabase db,
  int userId,
  String id, {
  String? name,
  String? type,
  List<String>? tagIds,
  Value<String?> optimizerConstraints = const Value.absent(),
  Value<String?> linkedModSetId = const Value.absent(),
  required String now,
}) async {
  final existing = await getSet(db, userId, id);
  if (existing == null) return null;

  await (db.update(db.sets)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .write(
    SetsCompanion(
      name: Value(name ?? existing.name),
      type: Value(type ?? existing.type),
      optimizerConstraints: optimizerConstraints.present
          ? optimizerConstraints
          : Value(existing.optimizerConstraints),
      linkedModSetId: linkedModSetId.present
          ? linkedModSetId
          : Value(existing.linkedModSetId),
      updatedAt: Value(now),
    ),
  );

  if (tagIds != null) {
    await (db.delete(db.setTags)..where((t) => t.setId.equals(id))).go();
    for (final tagId in tagIds) {
      await db
          .into(db.setTags)
          .insert(SetTagsCompanion.insert(setId: id, tagId: tagId));
    }
  }
  return getSet(db, userId, id);
}

/// Delete set if not attached. Throws [SetInUseException] when attachments exist.
///
/// Schema also enforces ON DELETE RESTRICT as a backstop.
Future<bool> deleteSetRecord(AppDatabase db, int userId, String id) async {
  final existing = await getSet(db, userId, id);
  if (existing == null) return false;

  final refs = await findAttachmentsBySetId(db, id);
  if (refs.isNotEmpty) {
    throw SetInUseException(id, refs);
  }

  final n = await (db.delete(db.sets)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .go();
  return n > 0;
}

/// True if another set already has (userId, type, name).
Future<bool> findDuplicateSetName(
  AppDatabase db,
  int userId, {
  required String type,
  required String name,
  String? excludeId,
}) async {
  final row = await (db.select(db.sets)
        ..where(
          (t) =>
              t.userId.equals(userId) &
              t.type.equals(type) &
              t.name.equals(name),
        ))
      .getSingleOrNull();
  if (row == null) return false;
  if (excludeId != null && row.id == excludeId) return false;
  return true;
}

/// Builds/variants that attach [setId] (for “in use” UX and RESTRICT checks).
Future<List<SetAttachmentRef>> findAttachmentsBySetId(
  AppDatabase db,
  String setId,
) async {
  final query = db.select(db.variantSetAttachments).join([
    innerJoin(
      db.buildVariants,
      db.buildVariants.id.equalsExp(db.variantSetAttachments.variantId),
    ),
    innerJoin(
      db.builds,
      db.builds.id.equalsExp(db.buildVariants.buildId),
    ),
  ])
    ..where(db.variantSetAttachments.setId.equals(setId));

  final rows = await query.get();
  return rows
      .map(
        (row) => SetAttachmentRef(
          buildId: row.readTable(db.builds).id,
          buildName: row.readTable(db.builds).name,
          variantId: row.readTable(db.buildVariants).id,
          variantName: row.readTable(db.buildVariants).name,
        ),
      )
      .toList();
}
